storage: clear the timestamp cache when we're not the lease holder

petermattis opened the pull request:
Fixes #9101.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9102)

<!-- Reviewable:end -->

petermattis committed:
storage: clear the timestamp cache when we're not the lease holder

Fixes #9101.

petermattis commented:
Cc @cockroachdb/stability

bdarnell commented:
LGTM

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica_trigger.go, line 324 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQsbKeVhYiSLl-_QUnA:-KQsbKeVhYiSLl-_QUnB:b-fht6mq) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/replica_trigger.go#L324)):*
> ```Go
> 			}
> 		} else if trigger.lease.Covers(r.store.Clock().Now()) {
> 			// We're not the lease holder, reset our timestamp cache, releasing
> ```

We should do this whether the lease covers the current time or not. 

---

*[storage/replica_trigger.go, line 330 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQscjXzc50JFb2xKIcL:-KQscjY-RmVsAhSNrcoq:bv1d6dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/replica_trigger.go#L330)):*
> ```Go
> 			r.mu.Lock()
> 			// TODO(tschottdorf): Why does calling TimestampCache.Clear() cause
> 			// TestReplicaTSCacheLowWaterOnLease to fail?
> ```

Which part of the test is failing? [This part](https://github.com/cockroachdb/cockroach/blob/62010bbf3d180dfbc490ab5dcf427d7c6ba4cc73/storage/replica_test.go#L916) appears to be testing that the low water mark is left alone when we lose the lease, so we should simply change the test to match the new behavior.

---

*[storage/timestamp_cache.go, line 154 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQsblPxmpULKOv9ti_X:-KQsblPyYFhEwIHVifr_:b-bk7rib) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/timestamp_cache.go#L154)):*
> ```Go
> 
> func (tc *timestampCache) reset() {
> 	tc.requests = btree.New(btreeDegree)
> ```

Remove the initialization of `tc.requests` in `newTimestampCache`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-:-KQscvFoAp-nuAplAWnZ:bb6ywmx)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/replica_trigger.go, line 324 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQsbKeVhYiSLl-_QUnA:-KQuK9OFHPvrcyO_qAbl:b-9ztdza) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/replica_trigger.go#L324)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We should do this whether the lease covers the current time or not. 
</details>
Is there ever a scenario that we're replaying on old request (and thus the lease isn't valid) while the current lease belongs to this replica? That doesn't seem possible as the new lease will be in a subsequent command we'll replay.

Changed.

---

*[storage/replica_trigger.go, line 330 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQscjXzc50JFb2xKIcL:-KQuN6rwlis-P9gN47Ho:bbrnqyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/replica_trigger.go#L330)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Which part of the test is failing? [This part](https://github.com/cockroachdb/cockroach/blob/62010bbf3d180dfbc490ab5dcf427d7c6ba4cc73/storage/replica_test.go#L916) appears to be testing that the low water mark is left alone when we lose the lease, so we should simply change the test to match the new behavior.
</details>
Changed the test. This does highlight that we're not just setting the low-water to `now`, but to `now + maxClockOffset`. Is that problematic?

---

*[storage/timestamp_cache.go, line 154 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQsblPxmpULKOv9ti_X:-KQuL8bxpupRy3q3IXnA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/timestamp_cache.go#L154)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Remove the initialization of `tc.requests` in `newTimestampCache`. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/replica_trigger.go, line 324 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQsbKeVhYiSLl-_QUnA:-KQuWRZPNlG6lMuKAryI:bynw9ad) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/replica_trigger.go#L324)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Is there ever a scenario that we're replaying on old request (and thus the lease isn't valid) while the current lease belongs to this replica? That doesn't seem possible as the new lease will be in a subsequent command we'll replay.
> 
> Changed.
</details>
If we were the lease holder, got partitioned away, another node got the lease, and then the lease became quiescent, then after the partition healed we'd still have data in our timestamp cache and the only indication we'd have that we should throw it away is replaying expired leases. (because we don't have an event for lease expiration)

Actually, this is a problem on its own - when a range lease becomes quiescent, we just stop refreshing it; the last holder of the lease will never clear its timestamp cache. Not a huge problem for now, but we should figure out how to trigger a cache clear when the lease expires at some point. 

---

*[storage/replica_trigger.go, line 330 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQscjXzc50JFb2xKIcL:-KQuXY4St69jU05Cbgym:b-nubaex) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/replica_trigger.go#L330)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Changed the test. This does highlight that we're not just setting the low-water to `now`, but to `now + maxClockOffset`. Is that problematic?
</details>
There's a `TODO(tschottdorf)` in `Clear()`, but for the purposes of this change I don't think it's a problem. We won't be able to get the lease again until at least `maxClockOffset` has elapsed, so we'll never see the low water mark moving backwards. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102)*
<!-- Sent from Reviewable.io -->

tamird commented on storage/replica_test.go:
```diff
@@ -878,7 +878,7 @@ func TestReplicaTSCacheLowWaterOnLease(t *testing.T) {
 	baseLowWater := baseRTS.WallTime
 
-	newLowWater := now.Add(50, 0).WallTime + baseLowWater
+	newLowWater := now.Add(0, 0).WallTime + baseLowWater + int64(maxClockOffset)
```
Remove this .Add?

tamird commented on storage/replica_trigger.go:
```diff
@@ -320,26 +320,38 @@ func (r *Replica) leasePostCommitTrigger(
-					// holder, then try to transfer the raft leadership to match the
-					// lease.
-					log.Infof(ctx, "range %v: replicaID %v transfer raft leadership to replicaID %v",
-						r.RangeID, replicaID, trigger.lease.Replica.ReplicaID)
```
Should this TODO have been removed?

tamird replied:
Eh this comment is anchored on the wrong line; I was referring to the TODO on Clear() below.

petermattis replied:
I removed the `TODO(tschottdorf)` I added in this PR. Not sure if you're referring to another one. Your comments don't seem to be anchored to the code correctly.

tamird replied:
Yeah, that's the one I was referring to.

On Sep 5, 2016 09:55, "Peter Mattis" <notifications@github.com> wrote:

> In storage/replica_trigger.go
> <https://github.com/cockroachdb/cockroach/pull/9102#discussion_r77525528>:
>
> > @@ -320,26 +320,38 @@ func (r *Replica) leasePostCommitTrigger(
> >  					r.gossipFirstRangeLocked(ctx)
> >  				}()
> >  			}
> > -		} else if trigger.lease.Covers(r.store.Clock().Now()) {
> > -			if err := r.withRaftGroup(func(raftGroup *raft.RawNode) error {
> > -				if raftGroup.Status().RaftState == raft.StateLeader {
> > -					// If this replica is the raft leader but it is not the new lease
> > -					// holder, then try to transfer the raft leadership to match the
> > -					// lease.
> > -					log.Infof(ctx, "range %v: replicaID %v transfer raft leadership to replicaID %v",
> > -						r.RangeID, replicaID, trigger.lease.Replica.ReplicaID)
>
> I removed the TODO(tschottdorf) I added in this PR. Not sure if you're
> referring to another one. Your comments don't seem to be anchored to the
> code correctly.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9102/files/9426b8ffc228132a577ac5513684b279aebe1eea#r77525528>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPDf3QdgTD9AZwWqUu3Ov7cA-k603ks5qnB9fgaJpZM4J0qfB>
> .
>

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica_test.go, line 881 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQuN7YxPG6Vb9XZgSd3-r2-881:-KQubmossOgbbdlf-1Hv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9426b8ffc228132a577ac5513684b279aebe1eea/storage/replica_test.go#L881)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Remove this .Add?
</details>
Done.

---

*[storage/replica_trigger.go, line 324 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102#-KQsbKeVhYiSLl-_QUnA:-KQubxypX3NLX9aHEiYJ:bzdni52) ([raw file](https://github.com/cockroachdb/cockroach/blob/838c3e695465a13f93fcaea1db2178e34415bcc6/storage/replica_trigger.go#L324)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> If we were the lease holder, got partitioned away, another node got the lease, and then the lease became quiescent, then after the partition healed we'd still have data in our timestamp cache and the only indication we'd have that we should throw it away is replaying expired leases. (because we don't have an event for lease expiration)
> 
> Actually, this is a problem on its own - when a range lease becomes quiescent, we just stop refreshing it; the last holder of the lease will never clear its timestamp cache. Not a huge problem for now, but we should figure out how to trigger a cache clear when the lease expires at some point. 
</details>
I'll file an issue about this. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9102)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM
